### PR TITLE
#1828 Resolve Cypress Test Issues

### LINKED ISSUE #1828: Fix Cypress Test case errors in Test harness
Ref: 3696

### PR DESCRIPTION
**Pull request by srikant-ch5**
Fixes #1828 

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by nmgokhale** [comment]
@srikant-ch5 Any estimate on when this PR will be ready for review? After this PR is merged, we can create a `13.0.0 `production ready version.

**Comment by srikant-ch5** [comment]
> Hi @tomlyn Canvas Cypress tests are failing for below components because there is mismatch between position of element and expected position for example for command `cy.verifyNodeTransform("Binding (entry) node", 300, 349.5);` Error is `expected 500 to be close to 300 +/- 2` .
> 
> * canvas/decorators.cy.js
> * canvas/drag.cy.js
> * canvas/links.cy.js
> * canvas/nodes-resize.cy.js
> * canvas/snap-to-grid.cy.js
> * canvas/undo-redo.cy.js
> 
> @nmgokhale I am looking into properties/properties-tooltips.cy.js test as we discussed its due to tooltip not showing up in tables in UI.

Hi @tomlyn Could you please let me know what needs to be done here in above test cases is it okay to update those numbers ?

File: canvas_modules/harness/cypress/e2e/canvas/clipboard.cy.js
Comment (nmgokhale) [inline canvas_modules/harness/cypress/e2e/canvas/clipboard.cy.js:557]: Instead of adding this everywhere, can you add this config in https://github.com/elyra-ai/canvas/blob/main/canvas_modules/harness/cypress/.eslintrc.js

**Comment by tomlyn** [comment]
Hi @srikant-ch5 I spent a lot of time looking at these issues in the afternoon and evening.

For decorators.cy.js -- this runs OK for me on my Mac except for this test: "Test positioning a decoration using distance on a straight line link" where I had to add cy.wait(10) before each verify command. I don't see the position problem with that file

For drag.cy.js -- I do see the position problem. I sent you a couple of videos in Slack. Unfortunately GitHub won't let me upload them here. 

For links.cy.js -- same this as with drag

I haven't looked at the other tests.

**Comment by srikant-ch5** [comment]
> Hi @srikant-ch5 I spent a lot of time looking at these issues in the afternoon and evening.
> 
> For decorators.cy.js -- this runs OK for me on my Mac except for this test: "Test positioning a decoration using distance on a straight line link" where I had to add cy.wait(10) before each verify command. I don't see the position problem with that file
> 
> For drag.cy.js -- I do see the position problem. I sent you a couple of videos in Slack. Unfortunately GitHub won't let me upload them here.
> 
> For links.cy.js -- same problem as with drag.cy.js
> 
> I haven't looked at the other tests.

Hi @tomlyn I have referred to canvas videos which you have shared to resolve below test cases and updated the PR.
canvas/decorators.cy.js
canvas/drag.cy.js
canvas/links.cy.js
canvas/nodes-resize.cy.js
canvas/snap-to-grid.cy.js

But I am still facing issues in canvas/undo-redo.cy.js .

@nmgokhale I have created a new issue https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3755 related to tooltip. So currently only two test cases are failing properties-tooltips.cy.js and undo-redo.cy.js

**Comment by nmgokhale** [comment]
@srikant-ch5 
> @nmgokhale I have created a new issue in abstract-canvas related to tooltip. So currently only two test cases are failing properties-tooltips.cy.js and undo-redo.cy.js

1. If the test in `properties-tooltips.cy.js` is failing because of tooltip issue, you can skip that test in this PR. And add a comment in abstract-canvas issue to enable that test after fixing the issue. 



2. Enable cypress tests in build using - Search for # TODO: Enable this while fixing issue - https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3696 in the code to enable cypress tests.

**Comment by tomlyn** [comment]
@srikant-ch5 I sent two files on Slack which should resolve the undoredo.cy.js problems.

**Comment by srikant-ch5** [comment]
> @srikant-ch5 I sent two files on Slack which should resolve the undoredo.cy.js problems.

Hi @tomlyn Thanks for the files. I have included those changes and now undo-redo test case is passing.
@nmgokhale I have skipped the failing tooltip test case. Now all the test cases are passing, could you please review this PR?

File: canvas_modules/harness/cypress/e2e/canvas/nodes-resize.cy.js
Comment (nmgokhale) [inline canvas_modules/harness/cypress/e2e/canvas/nodes-resize.cy.js:37]: There are 20 `TODO` comments in this PR. Remove TODO comments for all tests which are working.

Comment (srikant-ch5) [reply canvas_modules/harness/cypress/e2e/canvas/nodes-resize.cy.js:37]: Hi Neha I am keeping these TODO comments in the PR as we discussed.

**Review by nmgokhale — APPROVED** [review]

File: canvas_modules/harness/cypress/e2e/canvas/drag.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/drag.cy.js:70]: For each place where we added this comment please add another comment line underneath each one saying:
```
// See: https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3760
```

File: canvas_modules/harness/cypress/e2e/canvas/links.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/links.cy.js:700]: This doesn't look right. Looks like another instance where we need to click on the node before the drag.

File: canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js:91]: This doesn't look right. Needs fixing

File: canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js:103]: This also doesn't not look right.

File: canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js:205]: This look like another instance where the initial selection on the node doesn't work.  Please try replacing this 'moveNodeToPosition with the code we've used else where where we get the node and click on it.

File: canvas_modules/harness/cypress/e2e/canvas/nodes-resize.cy.js
Comment (tomlyn) [reply canvas_modules/harness/cypress/e2e/canvas/nodes-resize.cy.js:37]: @nmgokhale This is the issue described here https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3760
which we ought to fix if we can.  If we can fix it the tests would return to doing the automatic selection on drag.

File: canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:568]: Typo -- should be
TODO - Cannot verify an invalid entered value.

File: canvas_modules/harness/cypress/support/canvas/node-cmds.js
Comment (tomlyn) [reply canvas_modules/harness/cypress/support/canvas/node-cmds.js:218]: In other words, why is it necessary to wait for 0 seconds?

**Review by tomlyn — CHANGES_REQUESTED** [review]
I did the review. As part of this change do you think it would be possible to upgrade to the latest version of Cypress, after we get the test running to our satisfaction?

**Comment by srikant-ch5** [comment]
> I did the review. As part of this change do you think it would be possible to upgrade to the latest version of Cypress, after we get the test running to our satisfaction?

Hi @tomlyn I have upgraded Cypress to v13.7.3 and all the test cases are passing in my local. Also I have updated code as per your feedback please review.
Screenshot 2024-04-17 at 1 45 45 PM

File: canvas_modules/harness/cypress/e2e/canvas/drag.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/drag.cy.js:95]: Why is this clicking on the comment twice?

File: canvas_modules/harness/cypress/e2e/canvas/snap-to-grid.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/snap-to-grid.cy.js:120]: This should have the comments added elsewhere for the same workaround.

File: canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:59]: You missed the link to the issue here and for the change below.

File: canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:124]: Missed issue comment

File: canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js
Comment (tomlyn) [inline canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:563]: Please add a comment line here referencing the issue I opened for this problem which is: 
https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3759

**Review by tomlyn — CHANGES_REQUESTED** [review]
Hi Srikant - I competed my second review.

I noticed that there a lot of `cy.wait(10)` lines repeated many times after the same line of code. If we almost always have to add a wait after a function in the top level code, I think it would be better if the `cy.wait(10)` was added to the function being called rather than repeating it many times in the top level code. That way, the waits will not clutter up the code and also, if we write new testcases and reuse the function, we won't have to add in additional waits. Examples of what I mean are:
cy.clickOptionFromContextSubmenu
cy.clickColorFromContextSubmenu
cy.linkNodeOutputPortToNodeInputPort
You may be able to spot more ....

**Comment by srikant-ch5** [comment]
> Hi Srikant - I competed my second review.
> 
> I noticed that there a lot of `cy.wait(10)` lines repeated many times after the same line of code. If we almost always have to add a wait after a function in the top level code, I think it would be better if the `cy.wait(10)` was added to the function being called rather than repeating it many times in the top level code. That way, the waits will not clutter up the code and also, if we write new testcases and reuse the function, we won't have to add in additional waits. Examples of what I mean are: cy.clickOptionFromContextSubmenu cy.clickColorFromContextSubmenu cy.linkNodeOutputPortToNodeInputPort You may be able to spot more ....

Hi @tomlyn I have added cy.wait wherever test cases were failing but if we add wait in the commands then it would wait everytime its called which would slow down running time. For example there are 76 cy.wait's in total but if we add it in the command for verifyNumberOfLinks it would be called 98 times for one command and since there are many such commands it might take a bit longer to run cypress. Please let me know whether I should remove wait from top level and add those in commands. I have updated PR as per below feedback please review.

**Comment by tomlyn** [comment]
Hi @srikant-ch5 That's why I said "If we **almost always have to add a wait** after a function in the top level code, " so if you are happy that we're not calling them unnecessarily from the top level code then that is OK with me.  BTW even adding 98 wait(10) would be 980 milliseconds. So less than a second on tests that usually take about 30mins to run.  

BTW - I don't see the green check mark against my name in the Reviewers list so I am assuming you'll click the little recycle button when you want me and Neha to review again. (Sorry if you already know this, I just want to make sure we are on the same page)
image

**Comment by srikant-ch5** [comment]
> Hi @srikant-ch5 That's why I said "If we **almost always have to add a wait** after a function in the top level code, " so if you are happy that we're not calling them unnecessarily from the top level code then that is OK with me. BTW even adding 98 wait(10) would be 980 milliseconds. So less than a second on tests that usually take about 30mins to run.
> 
> BTW - I don't see the green check mark against my name in the Reviewers list so I am assuming you'll click the little recycle button when you want me and Neha to review again. (Sorry if you already know this, I just want to make sure we are on the same page) image

Hi @tomlyn we are calling wait only where its necessary so I hope this is fine for now. Please review once (I have raised review request again).

**Review by tomlyn — APPROVED** [review]

**Review by nmgokhale — APPROVED** [review]

**Pull request merged**
